In [1]:
import pandas as pd
import numpy as np

print("import successfully")

import successfully


In [2]:
customer_transaction_data = pd.read_csv(r"C:\Users\user\Downloads\FinLora_Customer_Transaction_Dataset.csv")

In [3]:
customer_transaction_data.head()

,transaction_id,customer_id,timestamp,home_country,source_currency,dest_currency,channel,amount_src,amount_usd,fee,...,ip_risk_score,kyc_tier,account_age_days,device_trust_score,chargeback_history_count,risk_score_internal,txn_velocity_1h,txn_velocity_24h,corridor_risk,is_fraud
0,fee8542d-8ee6-4b0d-9671-c294dd08ed26,402cccc9-28de-45b3-9af7-cc5302aa1f93,2022-10-03 18:40:59.468549+00:00,US,USD,CAD,ATM,278.19,278.19,4.25,...,0.123,standard,263,0.522,0,0.223,0,0,0.0,0
1,bfdb9fc1-27fe-4a85-b043-4d813d679259,67c2c6b3-ef0a-4777-a3f1-c84a851bb6ad,2022-10-03 20:39:38.468549+00:00,CA,CAD,MXN,web,208.51,154.29,4.24,...,0.569,standard,947,0.475,0,0.268,0,1,0.0,0
2,fc855034-3ea5-4993-9afa-b511d93fe5e8,6d0d9b27-fa26-45f8-93b1-2df29d182d9c,2022-10-03 23:02:43.468549+00:00,US,USD,CNY,mobile,160.33,160.33,2.70,...,0.437,enhanced,367,0.939,0,0.176,0,0,0.0,0
3,2cf8c08e-42ec-444d-a755-34b9a2a0a4ca,7bd5200c-5d19-44f0-9afe-8b339a05366b,2022-10-04 01:08:53.468549+00:00,US,USD,EUR,mobile,59.41,59.41,2.22,...,0.594,standard,147,0.551,0,0.391,0,0,0.0,0
4,d907a74d-b426-438d-97eb-dbe911aca91c,70a93d26-8e3a-4179-900c-a4a7a74d08e5,2022-10-04 09:35:03.468549+00:00,US,USD,INR,mobile,200.96,200.96,3.61,...,0.121,enhanced,257,0.894,0,0.257,0,0,0.0,0


In [4]:
customer_transaction_data.shape

(11400, 26)

In [5]:
customer_transaction_data.columns

Index(['transaction_id', 'customer_id', 'timestamp', 'home_country',
       'source_currency', 'dest_currency', 'channel', 'amount_src',
       'amount_usd', 'fee', 'exchange_rate_src_to_dest', 'device_id',
       'new_device', 'ip_address', 'ip_country', 'location_mismatch',
       'ip_risk_score', 'kyc_tier', 'account_age_days', 'device_trust_score',
       'chargeback_history_count', 'risk_score_internal', 'txn_velocity_1h',
       'txn_velocity_24h', 'corridor_risk', 'is_fraud'],
      dtype='object')

Checking for the missing values in eaach variable in the dataset

In [6]:
customer_transaction_data.isnull().sum()

transaction_id                 0
customer_id                    0
timestamp                     29
home_country                   0
source_currency                0
dest_currency                  0
channel                        0
amount_src                     0
amount_usd                   305
fee                          295
exchange_rate_src_to_dest      0
device_id                      0
new_device                     0
ip_address                   305
ip_country                   301
location_mismatch              0
ip_risk_score                  0
kyc_tier                     300
account_age_days               0
device_trust_score           295
chargeback_history_count       0
risk_score_internal            0
txn_velocity_1h                0
txn_velocity_24h               0
corridor_risk                  0
is_fraud                       0
dtype: int64

In [7]:
customer_transaction_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11400 entries, 0 to 11399
Data columns (total 26 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   transaction_id             11400 non-null  object 
 1   customer_id                11400 non-null  object 
 2   timestamp                  11371 non-null  object 
 3   home_country               11400 non-null  object 
 4   source_currency            11400 non-null  object 
 5   dest_currency              11400 non-null  object 
 6   channel                    11400 non-null  object 
 7   amount_src                 11400 non-null  object 
 8   amount_usd                 11095 non-null  float64
 9   fee                        11105 non-null  float64
 10  exchange_rate_src_to_dest  11400 non-null  float64
 11  device_id                  11400 non-null  object 
 12  new_device                 11400 non-null  bool   
 13  ip_address                 11095 non-null  obj

In [8]:
missing = customer_transaction_data.isnull().sum()

missing[missing > 0].sort_values(ascending=False)

amount_usd            305
ip_address            305
ip_country            301
kyc_tier              300
fee                   295
device_trust_score    295
timestamp              29
dtype: int64

Seven columns have missing values which include: amount_usd(305), ip_address(305), ip_country(301), kyc_tier(300), fee(295), device_trust_score(295), and timestamp(29). amount_src represents a transaction amount, which is expected to be numeric, but pandas has classified it as object, suggesting there may be invalid values, symbols, text, formatting issues, or corrupted entries. 

In [9]:

customer_transaction_data['amount_src'].value_counts(dropna=False).head(20)

amount_src
100.0     15
102.29     5
115.29     4
71.78      4
102.37     4
105.2      4
119.15     4
250.19     4
261.23     4
94.34      4
54.74      4
167.16     4
58.93      4
121.3      4
82.25      4
128.93     4
144.29     4
99.23      3
295.6      3
160.06     3
Name: count, dtype: int64

•	The dataset comprises 11,400 observations across 26 standardised variables. Initial data quality assessments identified several data type misalignments; specifically, the timestamp and amount_src variables are currently stored as generic object types. These will be converted to datetime and numeric types, respectively, to enable quantitative and time-series evaluation.
•	Furthermore, missing data were detected across several features, including amount_usd (305), ip_address (305), ip_country (301), kyc_tier (300), fee (295), device_trust_score (295), and timestamp (29). Appropriate imputation or removal strategies will be applied during preprocessing to preserve data reliability.
•	Finally, analysis of the target variable (is_fraud) reveals a severe class imbalance, with legitimate transactions accounting for 91% of the records and fraudulent transactions representing only 9%. To address the risk of algorithm bias toward the majority class and the unreliability of simple accuracy metrics, data resampling strategies and specialised evaluation criteria will be incorporated during model development.


In [10]:
# Checking for lebal imbalance
customer_transaction_data['is_fraud'].value_counts(normalize=True).reset_index()

,is_fraud,proportion
0,0,0.912544
1,1,0.087456


###DATA CLEANING AND PREPARATION
converting datatypes of time stampe to datetime

In [11]:
customer_transaction_data['timestamp']= pd.to_datetime(customer_transaction_data['timestamp'],errors='coerce')
#converting data type of income_usd to float
customer_transaction_data['amount_src']= pd.to_numeric(customer_transaction_data['amount_src'],errors='coerce')

In [12]:
# confirming the data type again
customer_transaction_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11400 entries, 0 to 11399
Data columns (total 26 columns):
 #   Column                     Non-Null Count  Dtype              
---  ------                     --------------  -----              
 0   transaction_id             11400 non-null  object             
 1   customer_id                11400 non-null  object             
 2   timestamp                  11339 non-null  datetime64[ns, UTC]
 3   home_country               11400 non-null  object             
 4   source_currency            11400 non-null  object             
 5   dest_currency              11400 non-null  object             
 6   channel                    11400 non-null  object             
 7   amount_src                 11396 non-null  float64            
 8   amount_usd                 11095 non-null  float64            
 9   fee                        11105 non-null  float64            
 10  exchange_rate_src_to_dest  11400 non-null  float64            
 11  de

To standardise data types, pd.to_datetime() and pd.to_numeric() were executed on timestamp and amount_src, utilising the errors='coerce' parameter to set invalid inputs to null. This explicit coercion surfaced underlying data corruption that had previously escaped detection as object types. Consequently, missing record counts rose from 29 to 61 for timestamp and from 0 to 4 for amount_src. These coerced missing values were systematically addressed during the broader data imputation and cleaning phase.

In [13]:
#checking for missing data again
customer_transaction_data.isnull().sum()


transaction_id                 0
customer_id                    0
timestamp                     61
home_country                   0
source_currency                0
dest_currency                  0
channel                        0
amount_src                     4
amount_usd                   305
fee                          295
exchange_rate_src_to_dest      0
device_id                      0
new_device                     0
ip_address                   305
ip_country                   301
location_mismatch              0
ip_risk_score                  0
kyc_tier                     300
account_age_days               0
device_trust_score           295
chargeback_history_count       0
risk_score_internal            0
txn_velocity_1h                0
txn_velocity_24h               0
corridor_risk                  0
is_fraud                       0
dtype: int64


•	amount_usd (305 missing): Imputed deterministically by multiplying amount_src by the corresponding currency exchange rate.

•	fee (295 missing): Imputed using conditional median strategy—filled by the median fee grouped by channel. Remaining missing values filled using overall median fee.

•	ip_country (301 missing): Imputed using the user's corresponding home_country.

•	kyc_tier (300 missing): Imputed using the global mode (most frequent value).

•	device_trust_score (295 missing): Imputed using conditional median strategy grouped by kyc_tier and new_device. Remaining missing values filled using overall median device_trust_score.


 timestamp variable which has 61 missing values and amount_src with 4 missing values will be dropped using the dropna method because the number of missing values in both vaiables is very small compared to the values on our dataset. The ip_address variable has missing values of 305 cannot be filled with any value as each ip_address is a unique number. It will be dropped using the dropna method.

In [14]:
#calculating the exchange rate per currency
exchange_rates = customer_transaction_data[customer_transaction_data['amount_usd'].notna()].groupby('source_currency').apply(
    lambda x: (x['amount_usd'] / x ['amount_src']).mean()
).to_dict()
exchange_rates

C:\Users\user\AppData\Local\Temp\ipykernel_51428\2926808893.py:2: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  exchange_rates = customer_transaction_data[customer_transaction_data['amount_usd'].notna()].groupby('source_currency').apply(


{'CAD': 0.7216095926871465,
 'GBP': 1.223441221648679,
 'USD': 0.9838730321259439}

In [15]:
#To fill the missing values of the amount_usd
customer_transaction_data['amount_usd']= customer_transaction_data.apply(
    lambda row: row['amount_usd'] if pd.notna(row['amount_usd']) else row['amount_src'] * exchange_rates.get(row['source_currency'],1),
    axis=1
)


In [16]:
# filling time missing vlues of the fee variable
if 'fee' in customer_transaction_data.columns:
    if 'channel' in customer_transaction_data:
        customer_transaction_data['fee'] = customer_transaction_data['fee'].fillna(customer_transaction_data['fee'].median())

In [17]:
# filling the missing valuesof ip_country variable
if {'ip_country', 'home_country'}.issubset(customer_transaction_data.columns):
    customer_transaction_data['ip_country'] = customer_transaction_data['ip_country'].fillna(customer_transaction_data['home_country'])

    

In [18]:
# to fill the missing values for device_trust_score variable
if 'device_trust_score' in customer_transaction_data.columns:
   if {'if new_device', 'kyc_tier'}.issubset(customer_transaction_data.columns):
    customer_transaction_data['device_trust_score']=customer_transaction_data.groupby(['new_device','kyc_tier',])['device_trust_score'].tramsform(lambda x: x.fillna(x.median()))
    customer_transaction_data['device_trust_score']=customer_transaction_data['device_trust_score'].fillna(customer_transaction_data['device_trust_score'].median())

In [19]:
#To drop the timestamp, amount_src and ip_address missing values
customer_transaction_data.dropna(subset=['timestamp', 'amount_src','ip_address'], inplace=True)

In [20]:
# checking for missing values again
customer_transaction_data.isnull().sum()

transaction_id               0
customer_id                  0
timestamp                    0
home_country                 0
source_currency              0
dest_currency                0
channel                      0
amount_src                   0
amount_usd                   0
fee                          0
exchange_rate_src_to_dest    0
device_id                    0
new_device                   0
ip_address                   0
ip_country                   0
location_mismatch            0
ip_risk_score                0
kyc_tier                     0
account_age_days             0
device_trust_score           0
chargeback_history_count     0
risk_score_internal          0
txn_velocity_1h              0
txn_velocity_24h             0
corridor_risk                0
is_fraud                     0
dtype: int64

###NORMALISATION

In [ ]:
# checking for negative values in numeric columns


In [22]:
# dropping the negative values in the numeric columns
customer_transaction_data = customer_transaction_data.drop(customer_transaction_data[
    (customer_transaction_data['amount_src'] < 0 ) |
    (customer_transaction_data['fee'] < 0) |
    (customer_transaction_data['device_trust_score'] < 0) |
    (customer_transaction_data['txn_velocity_1h'] < 0)].index)

    

In [23]:
# cross checking negative values

negative_counts = {
    'amount_src': (customer_transaction_data['amount_src'] < 0).sum(),
    'amount_usd': (customer_transaction_data['amount_usd']< 0).sum(),
    'fee': (customer_transaction_data['fee'] < 0).sum(),
    'device_trust_score':(customer_transaction_data['device_trust_score'] < 0).sum(),
    'txn_velocity_1h': (customer_transaction_data['txn_velocity_1h']< 0).sum(),
    'txt_velocity_24h':(customer_transaction_data['txn_velocity_24h'] < 0).sum(),
    'risk_score_internal': (customer_transaction_data['risk_score_internal'] < 0).sum(),
    }
negative_counts
    


{'amount_src': np.int64(0),
 'amount_usd': np.int64(0),
 'fee': np.int64(0),
 'device_trust_score': np.int64(0),
 'txn_velocity_1h': np.int64(0),
 'txt_velocity_24h': np.int64(0),
 'risk_score_internal': np.int64(0)}

In [24]:
# checking time stamp for futuristic values
customer_transaction_data[customer_transaction_data['timestamp'] > pd.Timestamp.utcnow()]

,transaction_id,customer_id,timestamp,home_country,source_currency,dest_currency,channel,amount_src,amount_usd,fee,...,ip_risk_score,kyc_tier,account_age_days,device_trust_score,chargeback_history_count,risk_score_internal,txn_velocity_1h,txn_velocity_24h,corridor_risk,is_fraud


In [25]:
#checking for location mismatch
customer_transaction_data['location_mismatch'].value_counts().reset_index()

,location_mismatch,count
0,False,9047
1,True,1793


In [26]:
# checking for channel mismatch
customer_transaction_data['channel']. unique()

array(['ATM', 'web', 'mobile', 'WEB', ' web  ', 'MOBILE', 'unknown',
       'mobille', ' mobile  ', 'weeb', 'ATm', ' ATM  '], dtype=object)

In [27]:
#channel variable formatting
customer_transaction_data['channel'] = customer_transaction_data['channel'].str.lower().str.strip()
customer_transaction_data['channel']. unique()


array(['atm', 'web', 'mobile', 'unknown', 'mobille', 'weeb'], dtype=object)

In [268]:
customer_transaction_data['channel'] = customer_transaction_data['channel']. replace({
    'weeb':'WEB',
    'unknown':np.nan,
    'web':'WEB',
    'mobille':'MOBILE',
    'ATm':'ATM',
    'mobile':'MOBILE',
})

In [325]:
customer_transaction_data['channel']= customer_transaction_data['channel'].str.lower().str.strip()
customer_transaction_data['channel'].unique()

array(['atm', 'web', 'mobile', nan], dtype=object)

In [337]:
customer_transaction_data['channel']= customer_transaction_data['channel'].replace('unknown',np.nan)
customer_transaction_data['channel'].unique()

array(['atm', 'web', 'mobile', nan], dtype=object)

In [338]:
# checking the source currency varriable
customer_transaction_data['source_currency'].unique()

array(['USD', 'CAD', 'GBP'], dtype=object)

In [339]:
#checking destination currency variable
customer_transaction_data['dest_currency'].unique()

array(['CAD', 'MXN', 'CNY', 'EUR', 'INR', 'GBP', 'PHP', 'NGN', 'USD'],
      dtype=object)

In [376]:
#checking the kyc_tier variable
customer_transaction_data['kyc_tier']. unique()

array(['standard', 'enhanced', 'low', nan, 'np.nan'], dtype=object)

In [377]:
# To format the channel variable
customer_transaction_data['kyc_tier'] = customer_transaction_data['kyc_tier'].str.lower().str.strip()
customer_transaction_data['kyc_tier'].unique()

array(['standard', 'enhanced', 'low', nan, 'np.nan'], dtype=object)

In [384]:
customer_transaction_data['kyc_tier'] =customer_transaction_data['kyc_tier'].replace({
    'standrd': 'standard',
    'enhancd': 'enhanced',
        'nan': np.nan,
        'np.nan': np.nan,
})
customer_transaction_data['kyc_tier'].unique()

array(['standard', 'enhanced', 'low', nan], dtype=object)

In [385]:
# reviewing the home country vriable
customer_transaction_data['home_country']=customer_transaction_data['home_country'].str.lower().str.strip()
customer_transaction_data['home_country'].unique()

array(['us', 'ca', 'uk', 'nan'], dtype=object)

In [386]:
customer_transaction_data['home_country'] = customer_transaction_data['home_country'].replace({
    'us': 'US',
    'ca': 'CA',
    'uk': 'UK',
    'unknown': 'nan'
})
customer_transaction_data['home_country'].unique()

array(['US', 'CA', 'UK', 'nan'], dtype=object)

In [387]:
customer_transaction_data['ip_country'] = customer_transaction_data['ip_country'].str.lower().str.strip()
customer_transaction_data['ip_country'].unique()

array(['us', 'ca', nan, 'uk', 'nan'], dtype=object)

In [388]:
customer_transaction_data['ip_country'] = customer_transaction_data['ip_country'].replace({
    'us': 'US',
    'ca': 'CA',
    'uk': 'UK',
    'unknown': 'nan',
    'nan': 'nan'
})
customer_transaction_data['ip_country'].unique()

array(['US', 'CA', nan, 'UK', 'nan'], dtype=object)

In [390]:
customer_transaction_data.dropna(inplace=True)

In [392]:
customer_transaction_data.to_csv(r'../Finlora_Dataset/artifacts/cleaned_Data.csv')